<a href="https://colab.research.google.com/github/liyueling13/Predicting-Banned-Books/blob/main/3)_Banned_Books_Add_Books_that_Aren't_Banned.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Add Books that Aren't Banned
11/7/23

Now let's do the same thing to another 1000 books -- that aren't banned.

## Setup

In [ ]:
# this performs line wrapping on output text in Colab

from IPython.display import HTML, display

def set_css():
  display(HTML('''
  <style>
    pre {
        white-space: pre-wrap;
    }
  </style>
  '''))
get_ipython().events.register('pre_run_cell', set_css)

In [ ]:
# mount drive
from google.colab import drive
drive.mount('/content/drive')
import os
os.chdir('/content/drive/My Drive/Data Science/Springboard assignments/Capstone Three/Banned Books')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd

## Get 1000 Most Read Books
I'm pulling from a private wordpress site: [Library of 1000 Books](https://libraryof1000books.wordpress.com/the-list-of-1000-books/). Scanning over the books, they seem like a reasonable list.

In [ ]:
from bs4 import BeautifulSoup, SoupStrainer
import requests

url = 'https://libraryof1000books.wordpress.com/the-list-of-1000-books/'

In [ ]:
r = requests.get(url)
text = r.text
print(text)

<!DOCTYPE html>
<html lang="en" class="no-js">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="profile" href="https://gmpg.org/xfn/11">
		<script>(function(html){html.className = html.className.replace(/\bno-js\b/,'js')})(document.documentElement);</script>
<title>The List of 1000 Books &#8211; Library of 1000 Books</title>
<script type="text/javascript">
  WebFontConfig = {"google":{"families":["Libre+Baskerville:r:latin,latin-ext","Vollkorn:r,i,b,bi:latin,latin-ext"]},"api_url":"https:\/\/fonts-api.wp.com\/css"};
  (function() {
    var wf = document.createElement('script');
    wf.src = 'https://s0.wp.com/wp-content/plugins/custom-fonts/js/webfont.js';
    wf.type = 'text/javascript';
    wf.async = 'true';
    var s = document.getElementsByTagName('script')[0];
    s.parentNode.insertBefore(wf, s);
	})();
</script><style id="jetpack-custom-fonts-css">.wf-active body, .wf-active button, .wf-active input, .wf-active sel

In [ ]:
soup = BeautifulSoup(text, 'html.parser')
books = soup.find('div', class_='entry-content')
books.text

'\nHere is the full List. All 1000 Books finally in one place\n\nThe Fault in Our Stars – John Green (2012)\nDivergent – Veronica Roth (2011)\nPride and Prejudice – Jane Austen (1813)\n1984 – George Orwell (1949)\nThe Kite Runner – Khaled Hosseini (2003)\nThe Fellowship of the Ring – J.R.R. Tolkien (1954)\nThe Lovely Bones – Alice Sebold (2002)\nGone Girl – Gillian Flynn (2012)\nThe Help – Kathyrn Stockett (2009)\nOf Mice and Men – John Steinbeck (1937)\nThe Alchemist – Paulo Coehlo (1988)\nFifty Shades of Grey – E.L. James (2011)\nThe Girl on the Train – Paula Hawkins (2015)\nThe Book Thief – Markus Zuzak (2005)\nLittle Women – Louise May Alcott (1862)\nEat, Pray, Love – Elizabeth Gilbert (2006)\nWater for Elephants – Sara Gruen (2006)\nThe Notebook – Nicholas Sparks (1996)\nLife of Pi – Yann Martel (2001)\nThe Hitchhiker’s Guide to the Galaxy – Douglas Adams (1979)\nWuthering Heights – Emily Bronte (1847)\nWhere the Sidewalk Ends – Shel Silverstein (1947)\nThe Perks of Being a Wallfl

In [ ]:
books_text = books.text
books_list = books_text.strip().split('\n')
books_list

['Here is the full List. All 1000 Books finally in one place',
 '',
 'The Fault in Our Stars – John Green (2012)',
 'Divergent – Veronica Roth (2011)',
 'Pride and Prejudice – Jane Austen (1813)',
 '1984 – George Orwell (1949)',
 'The Kite Runner – Khaled Hosseini (2003)',
 'The Fellowship of the Ring – J.R.R. Tolkien (1954)',
 'The Lovely Bones – Alice Sebold (2002)',
 'Gone Girl – Gillian Flynn (2012)',
 'The Help – Kathyrn Stockett (2009)',
 'Of Mice and Men – John Steinbeck (1937)',
 'The Alchemist – Paulo Coehlo (1988)',
 'Fifty Shades of Grey – E.L. James (2011)',
 'The Girl on the Train – Paula Hawkins (2015)',
 'The Book Thief – Markus Zuzak (2005)',
 'Little Women – Louise May Alcott (1862)',
 'Eat, Pray, Love – Elizabeth Gilbert (2006)',
 'Water for Elephants – Sara Gruen (2006)',
 'The Notebook – Nicholas Sparks (1996)',
 'Life of Pi – Yann Martel (2001)',
 'The Hitchhiker’s Guide to the Galaxy – Douglas Adams (1979)',
 'Wuthering Heights – Emily Bronte (1847)',
 'Where the 

In [ ]:
print(books_list[2])
print(books_list[1001])

# these are the first and last books of the 1000 book list

The Fault in Our Stars – John Green (2012)
A Moveable Feast – Ernest Hemingway (1964)


In [ ]:
books_list = books_list[2:1002]
len(books_list)

1000

In [ ]:
text = 'The Fault in Our Stars – John Green (2012)'
title, rest = map(str.strip, text.split('–'))
author, year = map(str.strip, rest.split('('))
year = year.rstrip(')')
title, author, year

('The Fault in Our Stars', 'John Green', '2012')

In [ ]:
import csv
file_path = 'thousand_books.csv'

with open(file_path, 'w', newline='', encoding='utf-8') as csvfile:
  csv_writer = csv.writer(csvfile)

  # Write header row
  csv_writer.writerow(['title', 'author', 'year'])

    # Iterate through the list of books, split title, author, and year, and write as rows
  for book in books_list:
        try:
          # Split the book information using '–' and '(' as delimiters
          title, author_and_year = map(str.strip, book.split('–'))
          author, year = map(str.strip, author_and_year.split('('))
          year = year.rstrip(')')  # Remove the closing parenthesis

          # Write title, author, and year as a row in the CSV file
          csv_writer.writerow([title, author, year])
        except:
          pass

print(f'Data has been written to {file_path}.')

Data has been written to thousand_books.csv.


In [ ]:
thousand_books = pd.read_csv('thousand_books.csv')

In [ ]:
thousand_books

author                       title  year
0                 John Green      The Fault in Our Stars  2012
1              Veronica Roth                   Divergent  2011
2                Jane Austen         Pride and Prejudice  1813
3              George Orwell                        1984  1949
4             J.R.R. Tolkien  The Fellowship of the Ring  1954
..                       ...                         ...   ...
961            Alan Brennert                    Moloka’i  2003
962            Joanne Harris                    Chocolat  1999
963  Hans Christian Anderson    The Complete Fairy Tales  1835
964       Tennessee Williams         The Glass Menagerie  1945
965         Ernest Hemingway            A Moveable Feast  1964

[966 rows x 3 columns]

In [ ]:
reordered = ['author', 'title', 'year']
thousand_books = thousand_books[reordered]
thousand_books

author                     title  year
0                 John Green    The Fault in Our Stars  2012
1              Veronica Roth                 Divergent  2011
2                Jane Austen       Pride and Prejudice  1813
3              George Orwell                      1984  1949
4            Khaled Hosseini           The Kite Runner  2003
..                       ...                       ...   ...
991            Alan Brennert                  Moloka’i  2003
992            Joanne Harris                  Chocolat  1999
993  Hans Christian Anderson  The Complete Fairy Tales  1835
994       Tennessee Williams       The Glass Menagerie  1945
995         Ernest Hemingway          A Moveable Feast  1964

[996 rows x 3 columns]

In [ ]:
thousand_books.to_csv('thousand_books.csv', index=False)

## Check for overlap
Are there any books in the thousand_books dataframe that are in the PEN America Banned Books dataframe?

In [ ]:
banned_df = pd.read_csv('banned_cleaned.csv')
banned_df

author                   title  \
0     Àbíké-Íyímídé, Faridah           Ace of Spades   
1         Acevedo, Elizabeth      Clap When You Land   
2         Acevedo, Elizabeth              The Poet X   
3                Acito, Marc  How I Paid for College   
4               Addasi, Maha            Time to Pray   
...                      ...                     ...   
1020                     Yin                 Coolies   
1021             Yolen, Jane               Encounter   
1022       Yousafzai, Malala   Malala's Magic Pencil   
1023         Zabarsky, Jessi              Witchlight   
1024        Zunon, Elizabeth           Grandpa Cacao   

                                    description_cleaned  
0     gossip girl meet ace spade thriller debut auth...  
1     grief love book award winning york time author...  
2     book award longlist title fan jacqueline woods...  
3     romp novel jersey teenager quest school tuitio...  
4     visit grandmother east yasmin time picture boo...  
...                                                 ...  
1020  shek marvel world brother wong california hund...  
1021     taino boy island salvador landing columbus men  
1022  peace prize winner york time author malala you...  
1023  love loss witch novel story story artwork lele...  
1024  story connects girl chocolate cake father gran...  

[1025 rows x 3 columns]

In [ ]:
overlap = thousand_books[thousand_books['title'].isin(banned_df['title'])]
overlap

# these thirty books are in both dataframes

author                            title  year
4      Khaled Hosseini                  The Kite Runner  2003
6         Alice Sebold                 The Lovely Bones  2002
7        Gillian Flynn                        Gone Girl  2012
9       John Steinbeck                  Of Mice and Men  1937
16          Sara Gruen              Water for Elephants  2006
22     Stephen Chbosky  The Perks of Being a Wallflower  1999
39     Jeannette Walls                 The Glass Castle  2005
49      Rainbow Rowell                   Eleanor & Park  2013
82       Lauren Oliver                    Before I Fall  2010
95   Jeffrey Eugenides                        Middlesex  2002
121       Sylvia Plath                     The Bell Jar  1963
159        Garth Stein    The Art of Racing in the Rain  2008
165  Katherin Paterson             Bridge to Terabithia  1977
266       Jodi Picoult                 Nineteen Minutes  2007
273      Toni Morrison                          Beloved  1987
348        J.M/ Barrie                        Peter Pan  1904
358          Anonymous                     Go Ask Alice  1971
368        Tim O’Brien          The Things They Carried  1990
376        Corrie Boom                 The Hiding Place  1971
393      Roy Arundhati          The God of Small Things  1997
463          Rupi Kaur                   Milk and Honey  2014
507       Angie Thomas                  The Hate U Give  2017
575          Yaa Gyasi                        Homegoing  2016
579     Isabel Allende         The House of the Spirits  1982
630        John Irving            The Cider House Rules  1985
684   Ta-Nehisi Coates         Between the World and Me  2015
695        Aziz Ansari                   Modern Romance  2015
718      Toni Morrison                   The Bluest Eye  1970
809       Ruta Sepetys           Between Shades of Gray  2011
907       Jodi Picoult                 The Tenth Circle  2006

In [ ]:
thousand_books.drop(overlap.index, inplace=True)

<ipython-input-76-2f301c469389>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  thousand_books.drop(overlap.index, inplace=True)


In [ ]:
thousand_books

author                       title  year
0                 John Green      The Fault in Our Stars  2012
1              Veronica Roth                   Divergent  2011
2                Jane Austen         Pride and Prejudice  1813
3              George Orwell                        1984  1949
5             J.R.R. Tolkien  The Fellowship of the Ring  1954
..                       ...                         ...   ...
991            Alan Brennert                    Moloka’i  2003
992            Joanne Harris                    Chocolat  1999
993  Hans Christian Anderson    The Complete Fairy Tales  1835
994       Tennessee Williams         The Glass Menagerie  1945
995         Ernest Hemingway            A Moveable Feast  1964

[966 rows x 3 columns]

In [ ]:
thousand_books.to_csv('thousand_books.csv', index=False)

Right now we have 1025 banned books, 966 non-banned books.

## Get descriptions

In [ ]:
from numpy import NaN
thousand_books['description'] = NaN
thousand_books

author                       title  year  description
0                 John Green      The Fault in Our Stars  2012          NaN
1              Veronica Roth                   Divergent  2011          NaN
2                Jane Austen         Pride and Prejudice  1813          NaN
3              George Orwell                        1984  1949          NaN
4             J.R.R. Tolkien  The Fellowship of the Ring  1954          NaN
..                       ...                         ...   ...          ...
961            Alan Brennert                    Moloka’i  2003          NaN
962            Joanne Harris                    Chocolat  1999          NaN
963  Hans Christian Anderson    The Complete Fairy Tales  1835          NaN
964       Tennessee Williams         The Glass Menagerie  1945          NaN
965         Ernest Hemingway            A Moveable Feast  1964          NaN

[966 rows x 4 columns]

In [ ]:
# API_KEY = redacted

In [ ]:
def paste_book_details(title, author):
    base_url = 'https://www.googleapis.com/books/v1/volumes'
    params = {
        'q': f'intitle:{title}+inauthor:{author}',
        'key': API_KEY
    }

    response = requests.get(base_url, params=params)

    if response.status_code == 200:
        data = response.json()
        if 'items' in data:
            book = data['items'][0]['volumeInfo']
            title = book.get('title', 'N/A')
            authors = ', '.join(book.get('authors', ['N/A']))
            description = book.get('description', 'N/A')
            thousand_books.loc[thousand_books['title']==title, 'description'] = description
        else:
            print(f'Title: {title} - Book not found.')
    else:
        print(f'Title: {title} - Error fetching data from Google Books API.')

In [ ]:
# test out code on the first 100
for i in range(0, 100):
    title = thousand_books.iloc[i]['title']
    author = thousand_books.iloc[i]['author']
    paste_book_details(title, author)

Title: The Help - Book not found.
Title: Tuesday with Morrie - Book not found.
Title: Lolita - Book not found.


In [ ]:
thousand_books[:100].isna().sum()

author          0
title           0
year            0
description    33
dtype: int64

In [ ]:
df_len = len(thousand_books)

import time
batch_size = 100

for i in range(100, df_len, batch_size):
    for j in range(i, min(i+batch_size, df_len)):
      title = thousand_books.iloc[j]['title']
      author = thousand_books.iloc[j]['author']
      paste_book_details(title, author)
    time.sleep(60)

Title: Inheart - Book not found.
Title: Orphan Train - Book not found.
Title: The Crucible - Book not found.
Title: Sill Alice - Book not found.
Title: The Thirteenth Tale - Book not found.
Title: Shatter Me - Book not found.
Title: Into the Water - Book not found.
Title: Pippy Longstocking - Book not found.
Title: Abrahan Lincoln: Vampire Hunter - Book not found.
Title: Naruto, Vol 1 - Book not found.
Title: Schingler’s List - Book not found.


In [ ]:
thousand_books.isna().sum()

author           0
title            0
year             0
description    273
dtype: int64

In [ ]:
# after fixing some titles and authors manually

thousand_books.isna().sum()

author           0
title            0
year             0
description    262
dtype: int64

In [ ]:
thousand_books.to_csv('thousand_books_descriptions.csv', index=False)

In [ ]:
thousand_books.dropna(inplace=True)

In [ ]:
thousand_books

author                   title  year  \
0            John Green  The Fault in Our Stars  2012   
1         Veronica Roth               Divergent  2011   
2           Jane Austen     Pride and Prejudice  1813   
5      Kathryn Stockett                The Help  2009   
6          Paulo Coehlo           The Alchemist  1988   
..                  ...                     ...   ...   
959          Jon Ronson     The Psychopath Test  2012   
960    Mark Danielewski         House of Leaves  2000   
962       Joanne Harris                Chocolat  1999   
964  Tennessee Williams     The Glass Menagerie  1945   
965    Ernest Hemingway        A Moveable Feast  1964   

                                           description  
0    Despite the medical miracle that has bought he...  
1    Paperback features over fifty pages of bonus m...  
2    Enter the world of Jane Austen's "Pride and Pr...  
5    The #1 New York Times bestselling novel and ba...  
6    A special 25th anniversary edition of the extr...  
..                                                 ...  
959  They say one out of every hundred people is a ...  
960  “A novelistic mosaic that simultaneously reads...  
962  Try me...Test me...Taste me.When an exotic str...  
964  Abandoned by her husband, Amanda Wingfield com...  
965  DigiCat Publishing presents to you this specia...  

[704 rows x 4 columns]

In [ ]:
# we have 1026 banned
# 704 non banned
# balanced enough?

## Data cleaning

Similarly with the other dataframe, let's:
*   Lowercase everything
*   Remove numbers and punctuation
*   Remove all words except nouns
*   Perform lemmatization (more computationally expensive, but preserves real words)
*   Remove random one letter words

In [ ]:
thousand_books['description_cleaned'] = thousand_books['description'].apply(lambda x: x.lower() if isinstance(x, str) else x)
thousand_books

author                   title  year  \
0            John Green  The Fault in Our Stars  2012   
1         Veronica Roth               Divergent  2011   
2           Jane Austen     Pride and Prejudice  1813   
5      Kathryn Stockett                The Help  2009   
6          Paulo Coehlo           The Alchemist  1988   
..                  ...                     ...   ...   
959          Jon Ronson     The Psychopath Test  2012   
960    Mark Danielewski         House of Leaves  2000   
962       Joanne Harris                Chocolat  1999   
964  Tennessee Williams     The Glass Menagerie  1945   
965    Ernest Hemingway        A Moveable Feast  1964   

                                           description  \
0    Despite the medical miracle that has bought he...   
1    Paperback features over fifty pages of bonus m...   
2    Enter the world of Jane Austen's "Pride and Pr...   
5    The #1 New York Times bestselling novel and ba...   
6    A special 25th anniversary edition of the extr...   
..                                                 ...   
959  They say one out of every hundred people is a ...   
960  “A novelistic mosaic that simultaneously reads...   
962  Try me...Test me...Taste me.When an exotic str...   
964  Abandoned by her husband, Amanda Wingfield com...   
965  DigiCat Publishing presents to you this specia...   

                                   description_cleaned  
0    despite the medical miracle that has bought he...  
1    paperback features over fifty pages of bonus m...  
2    enter the world of jane austen's "pride and pr...  
5    the #1 new york times bestselling novel and ba...  
6    a special 25th anniversary edition of the extr...  
..                                                 ...  
959  they say one out of every hundred people is a ...  
960  “a novelistic mosaic that simultaneously reads...  
962  try me...test me...taste me.when an exotic str...  
964  abandoned by her husband, amanda wingfield com...  
965  digicat publishing presents to you this specia...  

[704 rows x 5 columns]

In [ ]:
!pip install nltk
import nltk

nltk.download('punkt')

from nltk import word_tokenize, pos_tag
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [ ]:
def filter_nouns(text):
    words = word_tokenize(text)  # Tokenize the text into words
    tagged_words = pos_tag(words)  # Perform part-of-speech tagging

    # Keep only nouns (NN, NNS, NNP, NNPS)
    filtered_words = [word for word, tag in tagged_words if tag.startswith('NN')]

    filtered_text = ' '.join(filtered_words)  # Join the filtered words back into a string
    return filtered_text

In [ ]:
# get rid of all words that aren't nouns

%%time
thousand_books['description_cleaned'] = thousand_books['description_cleaned'].apply(lambda x: filter_nouns(x) if isinstance(x, str) else x)
thousand_books

CPU times: user 10.1 s, sys: 83.5 ms, total: 10.1 s
Wall time: 14.2 s


author                   title  year  \
0            John Green  The Fault in Our Stars  2012   
1         Veronica Roth               Divergent  2011   
2           Jane Austen     Pride and Prejudice  1813   
5      Kathryn Stockett                The Help  2009   
6          Paulo Coehlo           The Alchemist  1988   
..                  ...                     ...   ...   
959          Jon Ronson     The Psychopath Test  2012   
960    Mark Danielewski         House of Leaves  2000   
962       Joanne Harris                Chocolat  1999   
964  Tennessee Williams     The Glass Menagerie  1945   
965    Ernest Hemingway        A Moveable Feast  1964   

                                           description  \
0    Despite the medical miracle that has bought he...   
1    Paperback features over fifty pages of bonus m...   
2    Enter the world of Jane Austen's "Pride and Pr...   
5    The #1 New York Times bestselling novel and ba...   
6    A special 25th anniversary edition of the extr...   
..                                                 ...   
959  They say one out of every hundred people is a ...   
960  “A novelistic mosaic that simultaneously reads...   
962  Try me...Test me...Taste me.When an exotic str...   
964  Abandoned by her husband, Amanda Wingfield com...   
965  DigiCat Publishing presents to you this specia...   

                                   description_cleaned  
0    miracle years hazel anything augustus waters c...  
1    paperback pages bonus materials peek insurgent...  
2    world jane austen pride prejudice classic read...  
5    york times novel basis film—a timeless story l...  
6    edition bestseller foreword paulo coelho magic...  
..                                                 ...  
959  out people psychopath street today people empa...  
960  mosaic thriller excursion subconscious. ” —the...  
962  taste stranger vianne rocher village lansquene...  
964  husband wingfield comforts recollections life ...  
965  digicat publishing presents edition feast hemi...  

[704 rows x 5 columns]

In [ ]:
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
lemmatizer = WordNetLemmatizer()

In [ ]:
%%time
thousand_books['description_cleaned'] = thousand_books['description_cleaned'].apply(lambda x: ' '.join([lemmatizer.lemmatize(word) for word in x.split()]) if isinstance(x, str) else x)
thousand_books

CPU times: user 1.6 s, sys: 74.2 ms, total: 1.67 s
Wall time: 1.68 s


author                   title  year  \
0            John Green  The Fault in Our Stars  2012   
1         Veronica Roth               Divergent  2011   
2           Jane Austen     Pride and Prejudice  1813   
5      Kathryn Stockett                The Help  2009   
6          Paulo Coehlo           The Alchemist  1988   
..                  ...                     ...   ...   
959          Jon Ronson     The Psychopath Test  2012   
960    Mark Danielewski         House of Leaves  2000   
962       Joanne Harris                Chocolat  1999   
964  Tennessee Williams     The Glass Menagerie  1945   
965    Ernest Hemingway        A Moveable Feast  1964   

                                           description  \
0    Despite the medical miracle that has bought he...   
1    Paperback features over fifty pages of bonus m...   
2    Enter the world of Jane Austen's "Pride and Pr...   
5    The #1 New York Times bestselling novel and ba...   
6    A special 25th anniversary edition of the extr...   
..                                                 ...   
959  They say one out of every hundred people is a ...   
960  “A novelistic mosaic that simultaneously reads...   
962  Try me...Test me...Taste me.When an exotic str...   
964  Abandoned by her husband, Amanda Wingfield com...   
965  DigiCat Publishing presents to you this specia...   

                                   description_cleaned  
0    miracle year hazel anything augustus water can...  
1    paperback page bonus material peek insurgent a...  
2    world jane austen pride prejudice classic read...  
5    york time novel basis film—a timeless story li...  
6    edition bestseller foreword paulo coelho magic...  
..                                                 ...  
959  out people psychopath street today people empa...  
960  mosaic thriller excursion subconscious. ” —the...  
962  taste stranger vianne rocher village lansquene...  
964  husband wingfield comfort recollection life mo...  
965  digicat publishing present edition feast hemin...  

[704 rows x 5 columns]

In [ ]:
import re
thousand_books['description_cleaned'] = thousand_books['description_cleaned'].apply(lambda x: re.sub(r'[^\w\s]', ' ', x) if isinstance(x, str) else x)
thousand_books

author                   title  year  \
0            John Green  The Fault in Our Stars  2012   
1         Veronica Roth               Divergent  2011   
2           Jane Austen     Pride and Prejudice  1813   
5      Kathryn Stockett                The Help  2009   
6          Paulo Coehlo           The Alchemist  1988   
..                  ...                     ...   ...   
959          Jon Ronson     The Psychopath Test  2012   
960    Mark Danielewski         House of Leaves  2000   
962       Joanne Harris                Chocolat  1999   
964  Tennessee Williams     The Glass Menagerie  1945   
965    Ernest Hemingway        A Moveable Feast  1964   

                                           description  \
0    Despite the medical miracle that has bought he...   
1    Paperback features over fifty pages of bonus m...   
2    Enter the world of Jane Austen's "Pride and Pr...   
5    The #1 New York Times bestselling novel and ba...   
6    A special 25th anniversary edition of the extr...   
..                                                 ...   
959  They say one out of every hundred people is a ...   
960  “A novelistic mosaic that simultaneously reads...   
962  Try me...Test me...Taste me.When an exotic str...   
964  Abandoned by her husband, Amanda Wingfield com...   
965  DigiCat Publishing presents to you this specia...   

                                   description_cleaned  
0    miracle year hazel anything augustus water can...  
1    paperback page bonus material peek insurgent a...  
2    world jane austen pride prejudice classic read...  
5    york time novel basis film a timeless story li...  
6    edition bestseller foreword paulo coelho magic...  
..                                                 ...  
959  out people psychopath street today people empa...  
960  mosaic thriller excursion subconscious     the...  
962  taste stranger vianne rocher village lansquene...  
964  husband wingfield comfort recollection life mo...  
965  digicat publishing present edition feast hemin...  

[704 rows x 5 columns]

In [ ]:
def remove_one_letter(text):
  words = text.split()
  filtered_words = [word for word in words if len(word)>1]
  return ' '.join(filtered_words)

thousand_books['description_cleaned'] = thousand_books['description_cleaned'].apply(remove_one_letter)

## Finishing up


*   What are the most popular words in the thousand books dataframe cleaned descriptions?
*   What are the most popular words in both the banned and the thousand books cleaned descriptions?
*   Save dataframe



In [ ]:
from collections import Counter
thousand_counts = Counter()

for text in thousand_books['description_cleaned']:
  tokens = word_tokenize(text)
  thousand_counts.update(tokens)

thousand_counts.most_common(50)

[('time', 1082),
 ('life', 798),
 ('book', 758),
 ('york', 660),
 ('world', 646),
 ('novel', 634),
 ('story', 504),
 ('author', 458),
 ('series', 422),
 ('love', 404),
 ('year', 368),
 ('family', 330),
 ('woman', 272),
 ('way', 252),
 ('man', 244),
 ('reader', 242),
 ('war', 214),
 ('child', 214),
 ('day', 200),
 ('people', 196),
 ('work', 192),
 ('girl', 184),
 ('friend', 178),
 ('power', 166),
 ('edition', 164),
 ('heart', 162),
 ('tale', 160),
 ('character', 158),
 ('city', 156),
 ('everything', 154),
 ('history', 152),
 ('home', 152),
 ('death', 140),
 ('secret', 140),
 ('boy', 140),
 ('and', 140),
 ('century', 138),
 ('bestseller', 128),
 ('journey', 126),
 ('thing', 126),
 ('place', 124),
 ('fiction', 124),
 ('truth', 122),
 ('night', 118),
 ('father', 118),
 ('dark', 114),
 ('mother', 112),
 ('past', 112),
 ('jordan', 110),
 ('wheel', 110)]

In [ ]:
banned_df

author                   title  \
0     Àbíké-Íyímídé, Faridah           Ace of Spades   
1         Acevedo, Elizabeth      Clap When You Land   
2         Acevedo, Elizabeth              The Poet X   
3                Acito, Marc  How I Paid for College   
4               Addasi, Maha            Time to Pray   
...                      ...                     ...   
1020                     Yin                 Coolies   
1021             Yolen, Jane               Encounter   
1022       Yousafzai, Malala   Malala's Magic Pencil   
1023         Zabarsky, Jessi              Witchlight   
1024        Zunon, Elizabeth           Grandpa Cacao   

                                    description_cleaned  
0     gossip girl meet ace spade thriller debut auth...  
1     grief love book award winning york time author...  
2     book award longlist title fan jacqueline woods...  
3     romp novel jersey teenager quest school tuitio...  
4     visit grandmother east yasmin time picture boo...  
...                                                 ...  
1020  shek marvel world brother wong california hund...  
1021     taino boy island salvador landing columbus men  
1022  peace prize winner york time author malala you...  
1023  love loss witch novel story story artwork lele...  
1024  story connects girl chocolate cake father gran...  

[1025 rows x 3 columns]

In [ ]:
banned_counts = Counter()

for text in banned_df['description_cleaned']:
  tokens = word_tokenize(str(text))
  banned_counts.update(tokens)

banned_counts.most_common(50)

[('book', 834),
 ('story', 537),
 ('life', 530),
 ('time', 513),
 ('school', 453),
 ('author', 438),
 ('family', 432),
 ('girl', 405),
 ('friend', 381),
 ('love', 350),
 ('world', 314),
 ('york', 294),
 ('year', 287),
 ('boy', 277),
 ('people', 272),
 ('novel', 271),
 ('child', 270),
 ('way', 261),
 ('reader', 211),
 ('day', 194),
 ('thing', 179),
 ('mother', 179),
 ('father', 163),
 ('award', 158),
 ('review', 156),
 ('identity', 154),
 ('home', 151),
 ('parent', 145),
 ('gender', 144),
 ('woman', 143),
 ('everything', 132),
 ('friendship', 121),
 ('winner', 121),
 ('history', 121),
 ('community', 115),
 ('debut', 114),
 ('kid', 114),
 ('teen', 110),
 ('picture', 109),
 ('everyone', 109),
 ('student', 107),
 ('heart', 104),
 ('something', 101),
 ('truth', 100),
 ('relationship', 100),
 ('fan', 99),
 ('place', 98),
 ('sister', 94),
 ('secret', 90),
 ('word', 90)]

In [ ]:
thousand_dict = Counter(dict(thousand_counts.most_common(100)))
banned_dict = Counter(dict(banned_counts.most_common(100)))
thousand_dict & banned_dict

Counter({'time': 513,
         'life': 530,
         'book': 834,
         'york': 294,
         'world': 314,
         'novel': 271,
         'story': 537,
         'author': 438,
         'series': 43,
         'love': 350,
         'year': 287,
         'family': 432,
         'woman': 143,
         'way': 261,
         'man': 74,
         'reader': 211,
         'war': 52,
         'child': 270,
         'day': 194,
         'people': 272,
         'work': 71,
         'girl': 276,
         'friend': 267,
         'power': 79,
         'edition': 37,
         'heart': 104,
         'tale': 79,
         'character': 62,
         'city': 67,
         'everything': 132,
         'history': 121,
         'home': 151,
         'death': 52,
         'secret': 90,
         'boy': 210,
         'and': 79,
         'century': 35,
         'bestseller': 23,
         'journey': 82,
         'thing': 179,
         'place': 98,
         'fiction': 47,
         'truth': 100,
         'night': 77

In [ ]:
thousand_books_cleaned = thousand_books.drop(columns=['description', 'year'])
thousand_books_cleaned

author                   title  \
0            John Green  The Fault in Our Stars   
1         Veronica Roth               Divergent   
2           Jane Austen     Pride and Prejudice   
5      Kathryn Stockett                The Help   
6          Paulo Coehlo           The Alchemist   
..                  ...                     ...   
959          Jon Ronson     The Psychopath Test   
960    Mark Danielewski         House of Leaves   
962       Joanne Harris                Chocolat   
964  Tennessee Williams     The Glass Menagerie   
965    Ernest Hemingway        A Moveable Feast   

                                   description_cleaned  
0    miracle year hazel anything augustus water can...  
1    paperback page bonus material peek insurgent a...  
2    world jane austen pride prejudice classic read...  
5    york time novel basis film timeless story line...  
6    edition bestseller foreword paulo coelho magic...  
..                                                 ...  
959  out people psychopath street today people empa...  
960  mosaic thriller excursion subconscious the yor...  
962  taste stranger vianne rocher village lansquene...  
964  husband wingfield comfort recollection life mo...  
965  digicat publishing present edition feast hemin...  

[704 rows x 3 columns]

In [ ]:
thousand_books_cleaned.to_csv('thousand_cleaned.csv', index=False)